controllers/machine_controller_noderef.go

```diff
@@ -63,26 +59,65 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
 		return err
 	}
 
-	// Get the Node reference.
-	nodeRef, err := r.getNodeReference(remoteClient, providerID)
+	// Even if Status.NodeRef exists, continue to do the following checks to make sure Node is healthy
+	node, err := r.getNode(remoteClient, providerID)
 	if err != nil {
 		if err == ErrNodeNotFound {
+			// While a NodeRef is set in the status, getting ErrNodeNotFound error means the node is deleted.
+			// If Status.NodeRef is not set before, node still can be in the provisioning state.
+			if machine.Status.NodeRef != nil {
+				conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeNotFoundReason, clusterv1.ConditionSeverityError, "")
+				return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
+					"failed to find the Node owned by Machine %q in namespace %q", machine.Name, machine.Namespace)
+			}
+
+			logger.Error(err, "Failed to assign NodeRef")
+			r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
 			return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 20 * time.Second},
 				"cannot assign NodeRef to Machine %q in namespace %q, no matching Node", machine.Name, machine.Namespace)
 		}
-		logger.Error(err, "Failed to assign NodeRef")
-		r.recorder.Event(machine, apicorev1.EventTypeWarning, "FailedSetNodeRef", err.Error())
+		// This may be a transient client error and may not mean there is no matching Node.
 		return err
 	}
 
 	// Set the Machine NodeRef.
-	machine.Status.NodeRef = nodeRef
-	logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
-	r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &apicorev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
+			Name:       node.Name,
+			UID:        node.UID,
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
```

Comment:
When the machine NodeRef is just set in this reconcile, do we want to go ahead and set node related conditions below or just `return nil` like I did here and in the next reconcile, set them?

Reply:
Seems like there are no extra calls being made after this line of code, so I'd say you can remove the return and instead add a comment explaining why we're proceeding with the other operations?